### PR TITLE
Refax around Window measurement facility

### PR DESCRIPTION
**Pull request by ethouris**
Extracted longer and copy-paste-looking code into external functions. Potentially they can be reused in other measurement activities.

### CODE REVIEW

File: srtcore/utilities.h [hunk 1]
```diff
@@ -843,6 +843,126 @@ struct CallbackHolder
...
+        const int* para,
+        int& w_sum, IntCount& w_count, IntParaCount& w_paracount)
+{
+    IntCount count = 0;
+    int sum = 0;
+    IntParaCount parasum = 0;
+    for (; p != end; ++p, ++para)
+    {
+        // Throw away those that don't fit in the filter
+        if (*p < filter.lower || *p > filter.upper)
```
Comment (maxsharabayko) [inline srtcore/utilities.h:954]: To be identical with the previous version should be `<=` and `>=` here.

File: srtcore/window.cpp [hunk 2]
```diff
@@ -158,99 +158,44 @@ void srt::CPktTimeWindowTools::initializeWindowArrays(int* r_pktWindow, int* r_p
...
-   }
+    PassFilter<int> filter = GetPeakRange(window, replica, asize);
+
+    unsigned count = 0;
+    int sum = 0;
+
+    w_bytesps = 0;
+    unsigned long bytes = 0;
+    // // (explicit specialization due to problems on MSVC 2013 and 2015)
+    AccumulatePassFilterParallel<unsigned, unsigned long>(window, window + asize, filter, abytes,
```
Comment (maxsharabayko) [inline srtcore/window.cpp:171]: `window + asize` as an end pointer might trigger warnings on some static analyzers. Using just length might be better.
Just a thought.

File: srtcore/utilities.h [hunk 3]
```diff
@@ -843,6 +843,126 @@ struct CallbackHolder
...
+        v1 = p.first;
+        v2 = p.second;
+    }
+};
+
+template <class T1, class T2> inline
+PairProxy<T1, T2> Tie2(T1& v1, T2& v2)
+{
+    return PairProxy<T1, T2>(v1, v2);
+}
```
Comment (maxsharabayko) [inline srtcore/utilities.h:866]: 👍

File: srtcore/window.cpp [hunk 4]
```diff
@@ -158,99 +158,44 @@ void srt::CPktTimeWindowTools::initializeWindowArrays(int* r_pktWindow, int* r_p
...
-      }
-      ++ p;
-   }
+    PassFilter<int> filter = GetPeakRange(window, replica, psize);
+
+    int count;
+    int sum;
+    Tie2(sum, count) = AccumulatePassFilter(window, window + psize, filter);
+    count += 1;
+    sum += filter.median;
```
Comment (maxsharabayko) [inline srtcore/window.cpp:196]: Changing the order would make `count += 1` less confusing.
```suggestion
    sum += filter.median;
    count += 1;
```

File: srtcore/window.cpp (hunk 2 above)
Comment (ethouris) [reply srtcore/window.cpp:171]: Well, provided that this is taken as an array anyway (not as a generic container), it should work.

File: srtcore/utilities.h (hunk 3 above)
Comment (ethouris) [reply srtcore/utilities.h:866]: Yeah, nice thing, although C++11 has this already and C++17 provides even a better alternative.

File: srtcore/utilities.h (hunk 1 above)
Comment (ethouris) [reply srtcore/utilities.h:954]: Right! Forgot that De Morgan's laws apply also to relative operators.

**Review by maxsharabayko — CHANGES_REQUESTED** [review]
One comment remains unaddressed.

**Review by maxsharabayko — APPROVED** [review]

**Pull request merged**